src/s3path.jl

```diff
@@ -322,17 +322,23 @@ function FilePathsBase.sync(f::Function, src::AbstractPath, dst::S3Path; delete=
     end
 end
 
-function Base.readdir(fp::S3Path)
+function Base.readdir(fp::S3Path; join=false, sort=true)
     if isdir(fp)
         k = fp.key
         # Only list the files and "dirs" within this S3 "dir"
         objects = s3_list_objects(fp.config, fp.bucket, k; delimiter="")
 
         # Only list the basename and not the full key
-        basenames = unique!([s3_get_name(k, string(o["Key"])) for o in objects])
+        results = unique!([s3_get_name(k, string(o["Key"])) for o in objects])
 
-        # Lexographically sort the results
-        return sort!(filter!(!isempty, basenames))
+        # Filter out any empty object names which are valid in S3
+        filter!(!isempty, results)
+
+        # Sort results if sort=true
+        sort && sort!(results)
+
+        # Return results, possibly joined with the root path if join=true
+        return join ? joinpath.(fp, results) : results
```

Comment:
Hmm, one thing I just realized about this is that it introduces a type instability cause `join` determines whether we return strings or paths.

Reply:
Yeah, it does-- I've been using this and for me having the functionality has been well worth it, but it might depend on your use-case. Generally I find my code ends up like
```julia
for file in readdir(path; join=true)
    # some kind of filtering...
    endswith(string(file), ".arrow") || continue
    process_file!(results, file)
end
```
and the function-barrier (`process_file!`) means the instability is not an issue.

Reply:
I'm trying to remember, is there a reason you didn't want to use `readpath`? I'm wondering if for now I should just add those flags to both `readpath` and `readdir`, but just make the only difference that one return paths and one returns strings? In the future, I may make a breaking release which unifies the API and always returns paths for `readdir`.

Reply:
Ah, is that what `readpath` does? I didn't know that function exists and it's not really documented:
```julia
julia> using FilePathsBase

help?> readpath
search: readpath

  readpath(fp::P) where {P <: AbstractPath} -> Vector{P}
```

Usually I stick to the Base API since I might have a mix of S3Paths and Base (string) paths.

(To be clear, `join` doesn't just return S3Path's vs strings, but rather it determines or not it's the full path, including the bucket etc, or just the filename).

Reply:
> (To be clear, join doesn't just return S3Path's vs strings, but rather it determines or not it's the full path, including the bucket etc, or just the filename).

Yeah, I was just wondering if it would have made more sense to do `joinpath.(string(fp), results)` in the `readdir` function and left returning paths to `readpath`?

Reply:
Okay. I've opted away from the type instability in FilePathsBase since I have less control over how it'll be used. It also means I can support that keyword without introducing a breaking change. I may change `readdir` to always return a path type if a path type is given in the future... in which case I'll plan to update `S3Path` as well.

https://github.com/rofinn/FilePathsBase.jl/pull/152

Reply:
> I may change readdir to always return a path type if a path type is given

is there a way to do that while matching Base’s semantics? Like a RelativePath type or something?

Reply:
I've looked into the `RelativePath` idea, and it's very verbose. Without a proper traits based approach I don't think that's gonna happen anytime soon. Mostly likely I'd just wrap the resulting string in an appropriate path type and expect subtypes to decide what they want to do. I think there's also an issue where folks want `basename` to return a path as well.